# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/CacheConstants.h
# Commit: ffbb26d2a3c1
# Full Hash: ffbb26d2a3c186748ef9367f6d265cedffbe6a69
# Author: Morgan Rae Reschenberg <mreschenberg@mozilla.com>
# Date: 2022-08-02 09:41:16
# Regressor Bug: 1774043
# File Overlap Count: 1
# Description:
#   Bug 1774043: [Part 1] Add cache domains, caching logic for LABEL_FOR and LABELLED_BY relations r=Jamie
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D152100
# ==============================================================================

diff -r e0b787859230 -r ffbb26d2a3c1 accessible/base/CacheConstants.h
--- a/accessible/base/CacheConstants.h	Tue Aug 02 04:59:08 2022 +0300
+++ b/accessible/base/CacheConstants.h	Tue Aug 02 02:51:03 2022 +0000
@@ -7,6 +7,8 @@
 #ifndef _CacheConstants_h_
 #define _CacheConstants_h_
 
+#include "RelationType.h"
+
 namespace mozilla {
 namespace a11y {
 
@@ -28,6 +30,7 @@
   static constexpr uint64_t Spelling = ((uint64_t)0x1) << 13;
   static constexpr uint64_t Viewport = ((uint64_t)0x1) << 14;
   static constexpr uint64_t ARIA = ((uint64_t)0x1) << 15;
+  static constexpr uint64_t Relations = ((uint64_t)0x1) << 16;
   static constexpr uint64_t All = ~((uint64_t)0x0);
 };
 
@@ -43,6 +46,28 @@
   Update,
 };
 
+struct RelationData {
+  nsStaticAtom* const mAtom;
+  nsStaticAtom* const mValidTag;
+  RelationType mType;
+};
+
+/**
+ * This array of RelationData lists our relation types and the cache
+ * attribute atoms that store their targets. Attributes may describe
+ * different kinds of relations, depending on the element they originate on.
+ * For example, an <output> element's `for` attribute describes a CONTROLLER_FOR
+ * relation, while the `for` attribute of a <label> describes a LABEL_FOR
+ * relation.
+ * To ensure we process these attributes appropriately, RelationData.mValidTag
+ * contains the atom for the tag this attribute/relation type paring is valid
+ * on. If the pairing is valid for all tag types, this field is null.
+ */
+static constexpr RelationData kRelationTypeAtoms[] = {
+    {nsGkAtoms::aria_labelledby, nullptr, RelationType::LABELLED_BY},
+    {nsGkAtoms::_for, nsGkAtoms::label, RelationType::LABEL_FOR},
+};
+
 }  // namespace a11y
 }  // namespace mozilla
 